### ISSUE TITLE
Some ideas for kdb.h

### ISSUE THREAD

**Issue by domhof** [ElektraInitiative/libelektra#1332]
Feel free to comment!
In case the discussion gets longer, we can also split this up into several issues.

# KDB

1. KDB * kdbOpen (Key * errorKey)
    * In the docu the errorKey seems to be inconsistently referred as "parentKey".
https://doc.libelektra.org/api/current/html/group__kdb.html#ga6808defe5870f328dd17910aacbdc6ca

2. int kdbClose (KDB * handle, Key * errorKey)
    * kdbClose returns -1 if the passed in handle is NULL. The precondition already states that a valid handle must be passed, so passing NULL is a programmer error. Hence, the programmer won't check the return value anyway. Therefore we can remove the return type. The only reason I can think of where a user might want to pass NULL is when he doesn't want to check the returned handle of kdbOpen before calling kdbClose. But usually between those two functions other operations (kdbGet, kdbSet) would be performed which would fail before ever reaching kdbClose.

3. int kdbGet (KDB * handle, KeySet * ks, Key * parentKey)
int kdbSet (KDB * handle, KeySet * ks, Key * parentKey)
    * Why is the error-key again called parentKey? Furthermore, according to the documentation, parentKey is not only used to transmit error information but also to restrict the set-operation to keys below the parentKey in the hierarchy ("Then you promise to only modify or remove keys below this key."). Why does kdbSet not just store what's in ks?
    * In the most simple (probably also the most common) case, the user just wants to read the configuration KeySet for his application and do something with it. For this the user has to CREATE an empty KeySet himself before calling kdbGet while he actually wants to READ a KeySet. Furthermore the user could make a mistake by passing the wrong ks to kdbGet which might lead to hard to debug errors. Instead, kdbGet could just return a new KeySet. If the user wants to merge it with another existing KeySet he can do this by calling ksAppend afterwards.
    * When I first looked at the examples I found it irritating that "kdbSet" is so similar named to "keySet". From the name kdbSet a user cannot judge if the function persists anything to the store. Same goes for kdbGet: Does this impose any larger amount of work on the KDB (it does: reading the value from the store, accessing plugins, ...)?
I propose to rename them like this:
kdbGet -> kdbRead
kdbSet -> kdbWrite

4. I propose the following functions for KDB:
KDB * kdbOpen (Key * errorKey)
KDB * kdbClose (Key * errorKey)
KeySet * kdbRead (KDB * kdb, Key * errorKey) // Returned keySet has reference count of 1.
KeySet * kdbWrite (KDB * kdb, Key * errorKey)

# KeySet

1. KeySet * ksNew (size_t alloc, ...)
KeySet * ksVNew (size_t alloc, va_list va)
    * The documentation states that "either your first and only parameter is 0 or your last parameter must be KEY_END.". When you first look at example code like "KeySet * keys = ksNew (0, KS_END);" there is no obvious reason for the KS_END. Furthermore lists of "keyNew"s passed as parameters can become unreadable quite fast and you can easily forget the KS_END at the end. By exposing a function with variadic parameters we loose on type safety and compile time checks and gain very little in my opinion, so I would rather remove the function as a whole.
Instead I would introduce a function ksNew with the size as parameter and expose a KEYSET_DEFAULT_SIZE (or similar named). This way it is clear to the user what the parameter stands for and what the size will be (and also in case the default size changes in future Elektra versions he doesn't have to adapt the application's code).
#define KEYSET_DEFAULT_SIZE 16
KeySet * ksNew (size_t size)

2. int ksCopy (KeySet * dest, const KeySet * source)
    * Passing NULL as source (see documentation) is a "trick" that should be avoided in my opinion. The precondition should state that source must be a valid KeySet (it can be an empty KeySet though). Allowing NULL here can lead to hard to debug programming errors. If a user wants to clear a KeySet he should be using ksClear (which is missing in the documentation).
    * To be consistent with other functions in the API the order of the parameters should be inverted: source, dest. A good guideline in my opinion would be to always specify the parameter the function is applied on as the first parameter. In this case I want to copy the KeySet "source".
    * If dest and source point to the same KeySet the source would be cleared (in the current implementation).
    * I'm not sure if we need this function at all. I guess it's there for performance reasons. By reusing a otherwise not anymore needed KeySet as the destination, one can avoid the overhead of allocating memory for a new KeySet. But how often is it the case that one has such a KeySet, with the exact amount of memory allocated (too less would lead to an error, too much would "leak" memory).

3. KeySet * ksDup (const KeySet * source)
    * This can be renamed to ksCopy if we remove ksCopy (KeySet * dest, const KeySet * source).

4. int ksDel (KeySet * ks)
    * I think it might make sense to use reference counting for KeySets too (not only for Keys). Firstly, because in the future we might add features that hold references to KeySets, and secondly, the user needs to wrap his head only around one concept (reference counting) instead of two (reference counting + explicit delete) within the same API.

5. int ksNeedSync (const KeySet * ks)
    * This function is already marked as deprecated so we should remove it before v1.0.0.

6. ssize_t ksGetSize (const KeySet * ks)
    * As the other API functions also do not use "get" to read properties of KeySet (ksHead, ksTail, ...) I would omit it here too.

7. ssize_t ksAppend (KeySet * ks, const KeySet * toAppend)
    * For the same reason as stated in kdbClose, NULL should be forbidden as a precondition.

8. ssize_t ksAppendKey (KeySet * ks, Key * toAppend)
    * The documentation states "Will take ownership of the key toAppend. That means ksDel(ks) will remove the key unless the key: ...". Taking ownership means that the reference count of Key is increased. It should be >= 1 before already (keyNew should init it with 1), thus I would remove the part "That means ksDel(ks) will remove the key unless the key: ..." since this should not be the case.
    * For the same reason as stated in kdbClose, NULL should be forbidden as a precondition.
    * In the current implementation a comment states that the deletion of toAppend is "needed for ksAppendKey(ks, keyNew(0))". As a new user I would not expect that the passed in toAppend is deleted in any case. If ksNew sets the reference counter to 1 this is not needed anymore.

9. KeySet * ksCut (KeySet * ks, const Key * cutpoint)
    * For the same reason as stated in kdbClose, NULL should be forbidden as a precondition.
    * I find the name not explicit enough. "cut" could also mean "cut and throw away". Maybe ksExtract would be more suitable?
    * Like in ksLookupByName, could we 

10. Key * ksPop (KeySet * ks)
Key * ksNext (KeySet * ks)
Key * ksCurrent (const KeySet * ks)
Key * ksHead (const KeySet * ks)
Key * ksTail (const KeySet * ks)

    * For the same reason as stated in kdbClose, NULL should be forbidden as a precondition.
    * Maybe for a new user "*Key" (ksPopKey, ksNextKey, ..., ksAppendKey) would be more explicit.

11. int ksRewind (KeySet * ks)
    * Key * ksLookup (KeySet * ks, Key * key, option_t options)
    * Key * ksLookupByName (KeySet * ks, const char * name, option_t options)
    * For the same reason as stated in kdbClose, NULL should be forbidden as a precondition.

# Key

1. Key * keyNew (const char * name, ...)
Key * keyVNew (const char * name, va_list va)
Key * keyDup (const Key * source)
int keyCopy (Key * dest, const Key * source)

    * Same as for KeySets
    * The reference counter should start with 1.

2. ssize_t keyIncRef (Key * key)

    * I would rename this to "keyRetain" because it expresses semantically better what it does (it's an implementation detail that a reference counter is incremented).

3. ssize_t keyDecRef (Key * key)

    * I would rename this to "keyRelease" because it expresses semantically better what it does (it's an implementation detail that a reference counter is decreased).

4. int keyIsBinary (const Key * key)
int keyIsString (const Key * key)

    * For the same reason as stated in kdbClose, NULL should be forbidden as a precondition.

    * "Is" indicates that it returns a boolean value while in fact it returns a 1, 0 or -1 (for NULL pointer). In an if () the -1 would be unexpectedly treated as true.

5. ssize_t keyGetString (const Key * key, char * returnedString, size_t maxSize)
ssize_t keySetString (Key * key, const char * newStringValue)
ssize_t keyGetBinary (const Key * key, void * returnedBinary, size_t maxSize)
ssize_t keySetBinary (Key * key, const void * newBinary, size_t dataSize)

    * For the same reason as stated in kdbClose, NULL should be forbidden as a precondition.

6. int keyDel (Key * key)

    * Can be removed if we set the reference counter to 1 in keyNew and free delete the key in keyRelease if the counter is 0.

7. ssize_t keyGetRef (const Key * key)

    * Should be marked as to be used for debugging only.

**Comment by markus2330** [comment]
# KDB

> parentKey

The basic idea was: it is called errorKey if the key is exclusively used for errors+warnings, but not to indicate parent information.

> `kdbClose` The precondition already states that a valid handle must be passed

For end user APIs you also need to check preconditions. And doing nothing silently might also not be the best option?

> Therefore we can remove the return type.

No, `kdbClose` actually can fail if plugins are not able to close themselves.

> Why does kdbSet not just store what's in ks?

There might be much more in ks as one wants to write back.

> kdbRead, kdbWrite

What about kdbFetch and kdbStore?

> `KeySet * kdbWrite (KDB * kdb, Key * errorKey)`

How should this work?

# KeySet

> variadic parameters we loose on type safety and compile time checks and gain very little

We would not lose compile time checks if we use a `printf`-format like specifier.

We mainly gain a way to completely specify configuration with a single expression which can be `#include`d or returned from macros. This feature is heavily used to have built-in configuration and for unit tests.

> KS_END

We changed it that you should always use `KS_END`, we even added a compile time check for it.

> alloc parameter

We should do a benchmark if a alloc parameter is useful at all.

(stopped for today before ksCopy)

Maybe you really should split up the issue! (You can also move my KeySet part to another issue.) Btw. the most urgent part (kdbhl for lcdproc) is missing! I am looking forward to such good feedback there as well! (please also in a separate issue)

(to be continued...)

**Comment by markus2330** [comment]
> Passing NULL as source (see documentation) is a "trick" that should be avoided in my opinion.

Yes, I agree. But this can also be handled in a higher-level API.

> If a user wants to clear a KeySet he should be using ksClear (which is missing in the documentation).

We optimized such functions away to reduce the library size of the core. They should be introduced at a higher level!

> To be consistent with other functions in the API the order of the parameters should be inverted: source, dest.

Can you point to the inconsistency? dest is always first?

> I'm not sure if we need this function at all.

You are right, it might not be as useful anymore. In earlier versions one could also clean/copy keys that are within a keyset. As this is not possible anymore, the need to clean/copy keys is reduced.

> the user needs to wrap his head only around one concept (reference counting) instead of two (reference counting + explicit delete) within the same API.

Yes, this is very true and a very good proposal! It also makes the use of bindings more consistent, for example, in C++ binding constructing a Key out of a Key is done by ref counting while KeySets get duplicated. This inconsistency is really annoying.

> This function is already marked as deprecated so we should remove it before v1.0.0.

Exactly!

> As the other API functions also do not use "get" to read properties of KeySet (ksHead, ksTail, ...) I would omit it here too.

Yes, very good idea.

> The documentation states

Please create a PR or separate issue for docu issues.

> In the current implementation a comment states that the deletion of toAppend is "needed for ksAppendKey(ks, keyNew(0))". As a new user I would not expect that the passed in toAppend is deleted in any case. If ksNew sets the reference counter to 1 this is not needed anymore.

Please elaborate.  We remove the key, not the keyset, why would a reference counting of ksNew change the situation?

> I find the name not explicit enough. "cut" could also mean "cut and throw away". Maybe ksExtract would be more suitable?

We found out the function is not generic as it should be anyway. We can replace it with a `ksFilter` where the user can pass a function how it should be filtered.

> Like in ksLookupByName, could we

Sentence stops in the middle?

> ksRewind

You did not comment if you like the internal cursor at all?

> The reference counter should start with 1.

It would be an implementation detail anyway if we remove `keyGetRef`.

> keyRetain, keyRelease

Good suggestions!

> keyIsBinary

You did not comment on the binary handling in general. What do you think about the use of the meta data binary?

> Can be removed if we set the reference counter to 1 in keyNew and free delete the key in keyRelease if the counter is 0.

Can you point to an API where it is done that way?

> Should be marked as to be used for debugging only.

We could create a debugging library which allows (read-only?) access to internals such as how much is currently allocated, the synchronization bits, and ref counting.

Thank you, the feedback was very productive! I am looking forward to comments on the high-level API.

**Comment by kodebach** [comment]
@markus2330 Is any of this still relevant? If so, @lawli3t should probably take a look.

**Comment by lawli3t** [comment]
I'll take a look on it and see if there is some stuff in there not covered by the respective API reviews1

**Comment by markus2330** [comment]
@lawli3t thank you! Make sure to look at it only after the reviews to not influence the outcome.

**Comment by markus2330** [comment]
@kodebach maybe you can take a look, too?

**Comment by kodebach** [comment]
I don't really think there is anything here that is still relevant and not covered somewhere else. The only thing I noticed, was renaming `*IncRef`/`*DecRef` to `*Retain`/`*Release`. We could do it, but since our reference counting mechanism is not exactly standard I think it is good that the name explicitly refers to references.

**Comment by markus2330** [comment]
Thank you for checking. I agree, we can close this then.

@domhof thank you a lot for the input. We look forward to another review by you :sparkling_heart:
